dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java

```diff
@@ -1278,38 +1265,41 @@ private void inlineLate(Function function) {
-        final Program program = call.getFunction().getProgram();
-        Register register = ((ValueFunctionCall) call).getResultRegister();
-        String name = call.getCalledFunction().getName();
+        final Register result = getResultRegister(call);
+        final String name = call.getCalledFunction().getName();
         final String separator = "nondet_";
-        int index = name.indexOf(separator);
+        final int index = name.indexOf(separator);
         assert index > -1;
-        String suffix = name.substring(index + separator.length());
+        final String suffix = name.substring(index + separator.length());
 
-        // Nondeterministic booleans
+        final Type nonDetType;
+        final boolean signed;
         if (suffix.equals("bool")) {
-            final Expression value = program.newConstant(types.getBooleanType());
-            final Expression cast = expressions.makeCast(value, register.getType());
-            return List.of(EventFactory.newLocal(register, cast));
-        }
-
-        // Nondeterministic integers
-        boolean signed = switch (suffix) {
-            case "int", "short", "long", "char" -> true;
-            default -> false;
-        };
+            // Nondeterministic booleans
+            signed = false;
+            nonDetType = types.getBooleanType();
+        } else {
+            // Nondeterministic integers
+            final int bits = switch (suffix) {
+                case "long", "ulong" -> 64;
+                case "int", "uint", "unsigned_int" -> 32;
+                case "short", "ushort", "unsigned_short" -> 16;
+                case "char", "uchar" -> 8;
+                default -> throw new UnsupportedOperationException(String.format("%s is not supported", call));
+            };
 
-        final int bits = switch (suffix) {
-            case "long", "ulong" -> 64;
-            case "int", "uint", "unsigned_int" -> 32;
-            case "short", "ushort", "unsigned_short" -> 16;
-            case "char", "uchar" -> 8;
-            default -> throw new UnsupportedOperationException(String.format("%s is not supported", call));
-        };
+            signed = switch (suffix) {
+                case "int", "short", "long", "char" -> true;
+                default -> false;
+            };
+            nonDetType = types.getIntegerType(bits);
+        }
 
-        final NonDetValue value = (NonDetValue) call.getFunction().getProgram().newConstant(types.getIntegerType(bits));
-        value.setIsSigned(signed);
-        return List.of(EventFactory.newLocal(register, expressions.makeCast(value, register.getType(), signed)));
+        final Register nonDetReg = call.getFunction().getOrNewRegister("__r_nondet_" + suffix, nonDetType);
+        return List.of(
+                EventFactory.Svcomp.newSignedNonDetChoice(nonDetReg, signed),
+                EventFactory.newLocal(result, expressions.makeCast(nonDetReg, result.getType(), signed))
```

Comment:
Do we really need this cast?

Reply:
Definitely for nondeterministic booleans, because the register will be `bv8` instead of `bv1/bool`. Apart from that case, I don't think it is needed.